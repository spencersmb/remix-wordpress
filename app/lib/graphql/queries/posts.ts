import gql from 'graphql-tag';
 
export const QUERY_NEXT_POSTS = gql`
    query NextPosts($after: String) {
        posts(first: 10, after: $after) {
            __typename
            pageInfo{
                endCursor
                hasNextPage
                hasPreviousPage
                startCursor
                __typename
            }
            edges {
                __typename
                node {
                    __typename
                    author {
                        node {
                            avatar {
                                height
                                url
                                width
                            }
                            id
                            name
                            slug
                        }
                    }
                    id
                    categories {
                        edges {
                            node {
                                databaseId
                                id
                                name
                                slug
                            }
                        }
                    }
                    content
                    date
                    excerpt
                    featuredImage {
                        node {
                            altText
                            caption
                            sourceUrl
                            srcSet
                            sizes
                            id
                        }
                    }
                    modified
                    databaseId
                    title
                    slug
                    isSticky
                }
            }
        }
        allSettings {
            readingSettingsPostsPerPage
        }
    }
`;

export const POST_FEATURED_IMAGE = gql`
  fragment featuredImageFields on Post {
    featuredImage {
      node {
        mediaDetails {
          width
          height
          sizes{
            width
            file
            height
            name
            sourceUrl
            mimeType
          }
        }
          altText
          caption
          sourceUrl
          srcSet
          sizes
          id
        }
      }
    }
`
export const RESOURCE_FEATURED_IMAGE = gql`
  fragment featuredImageFields__RL on ResourceLibrary {
    featuredImage {
      node {
        mediaDetails {
          sizes{
            width
            file
            height
            name
            sourceUrl
            mimeType
          }
        }
          altText
          caption
          sourceUrl
          srcSet
          sizes
          id
        }
      }
    }
`
export const RELEATED_POSTS_FIELDS = gql`
  fragment relatedPostsFields on Post {
    relatedPosts {
      ...featuredImageFields
      title
      slug
      tutorialManager {
       thumbnail {
        type
        image {
          altText
          caption
          sourceUrl
          srcSet
          sizes
          id
          mediaDetails{
            sizes{
              width
              file
              height
              name
              sourceUrl
              mimeType
            }
          }
        }
       }
      }
      categories {
        edges {
          node {
            slug
            name
          }
        }
      }
    }
  }
`
export const POST_BASIC_FIELDS = gql`
  fragment postBasicFields on Post {
    id
    content
    date
    dateGmt
    excerpt
    modified
    databaseId
    title
    slug
    isSticky
    categories {
      edges {
          node {
            databaseId
            id
            name
            slug
          }
      }
    }
    tags{
      edges{
          node{
            name
            slug
          }
      }
    }
  }
`
export const POST_RESOURCE_FIELDS = gql`
  fragment postResourceFields on Post_Tutorialmanager {
    resources {
    ... on Post_Tutorialmanager_Resources_PaidProduct {
      product {
        ... on Product {
          title
          slug
          featuredImage {
            node {
              mimeType
              mediaDetails {
                height
                width
                sizes{
                  width
                  file
                  height
                  name
                  sourceUrl
                  mimeType
                }
              }
                altText
                caption
                sourceUrl
                srcSet
                sizes
                id
            }
          }
          productDetails {
            licences {
              licenseType
              price
              url
            }
          }
        }
      }
      description
    }
    ... on Post_Tutorialmanager_Resources_Course {
      course {
        ... on Course {
          title
          slug
          link
          details {
            courseUrl
          }
          featuredImage {
            node {
              mimeType
              mediaDetails {
                height
                width
                sizes{
                  width
                  file
                  height
                  name
                  sourceUrl
                  mimeType
                }
              }
                altText
                caption
                sourceUrl
                srcSet
                sizes
                id
            }
          }
        }
      }
      description
    }
    ... on Post_Tutorialmanager_Resources_Download {
      download {
        name
        description
        url
      }
    }
    ... on Post_Tutorialmanager_Resources_ColorSwatch {
      colorSwatch {
        url
      }
    }
  }
}
`
export const PRODUCT_FIELDS = gql`
  fragment productFields on Product {
    title 
    slug
    featuredImage {
      node {
        mimeType
        mediaDetails {
          height
          width
          sizes{
            width
            file
            height
            name
            sourceUrl
            mimeType
          }
        }
          altText
          caption
          sourceUrl
          srcSet
          sizes
          id
      }
    }
    productDetails{
      licences {
        licenseType
        price
        url
      }
    }
  }
`
export const QUERY_POST_BY_ID = gql`
    ${POST_BASIC_FIELDS}
    ${POST_FEATURED_IMAGE}
    ${RELEATED_POSTS_FIELDS}
    query postById($id: ID!) {
        post(idType: DATABASE_ID, id: $id) {
          ...postBasicFields
          ...featuredImageFields
          ...relatedPostsFields
          seo{
              title
              opengraphPublishedTime
              opengraphModifiedTime
              metaDesc
              readingTime
          }
          author {
              node {
                  avatar {
                      height
                      url
                      width
                  }
                  id
                  name
                  slug
                  uri
              }
          }
          etSocialNav{
              pinterestMeta {
                description
              }
              pinterestImage{
                altText
                caption
                sourceUrl
                srcSet
                sizes
                id
                mediaDetails{
                  sizes{
                    width
                    file
                    height
                    name
                    sourceUrl
                    mimeType
                  }
                }
              }
          }
          tutorialManager {
              postExcerpt
              thumbnail {
                image {
                  altText
                  caption
                  sourceUrl
                  srcSet
                  sizes
                  id
                  mediaDetails{
                    width
                    height
                    sizes{
                      width
                      file
                      height
                      name
                      sourceUrl
                      mimeType
                    }
                  }
                }
              }
              youtube {
                id
                addVideoMetadata
                duration
                videoObject {
                  description
                  uploadDate
                  thumbnailUrl
                  clipElements {
                    name
                    startOffset
                    endOffset
                  }
                  potentialActions{
                    name
                    startOffset
                  }
                }
              }
              downloads {
                  ... on ResourceLibrary {
                      title
                      freebie {
                          downloadLink
                      }
                  }
              }
          }
          comments(first: 500, after: null, where: {parent: null}) {
              pageInfo{
                endCursor
                hasNextPage
              }
              edges {
                  node {
                      databaseId
                      approved
                      parent {
                        node {
                          databaseId
                        }
                      }
                      id
                      author {
                          node {
                            isRestricted
                              name
                              ... on CommentAuthor {
                                gravatar{
                                  url
                                }
                              }
                          }
                      }
                      date
                      commentedOn {
                          node {
                              id
                              status
                          }
                      }
                      content
                      replies {
                          edges {
                              node {
                                  id
                                  databaseId
                                  content
                                  date
                                  parent {
                                    node {
                                      databaseId
                                    }
                                  }
                                  author {
                                      node {
                                          name
                                        ... on CommentAuthor {
                                          gravatar{
                                            url

                                          }
                                        }
                                      }
                                  }
                                  
                              }
                          }
                      }
                  }
              }
          }
        }
    }
  `