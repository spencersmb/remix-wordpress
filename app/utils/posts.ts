import { isEmpty } from "lodash";

export function flattenAllPosts(posts:any): IPost[] | false{
  const postsFiltered = posts?.edges?.map(({ node = {} }) => node);
  return Array.isArray(postsFiltered) && postsFiltered.map(mapPostData)
}

export function rearrangeLicenses(licenses: ILicense[]){
  return licenses.reduce((acc: any, licence) => {
        // Add it to the beginning of the array
        if(licence.licenseType === 'standard'){
          acc.unshift(licence)
        }
        console.log('licence.licenseType', licence.licenseType);
        
        // if there is only one item in array, add extended as the 2nd
        if(licence.licenseType === 'extended'){
          acc.push(licence)
        }

        // if the ext is the last item to get looped over
        if(licence.licenseType === 'extended' && acc.length === 2){
          const standard = acc[0]
          const server = acc[1]
          acc = [
            standard,
            licence,
            server
          ]
        }

        if(licence.licenseType === 'server'){
          acc.push(licence)
        }

        return acc
      }, [])
}

export function mapPostData(post:IPostRaw | {} = {}): IPost {
  const data = { ...post };
  let modifiedData: any = {...post}

  // Clean up the author object to avoid someone having to look an extra
  // level deeper into the node
  if (data.author)  {
    modifiedData.author = {
      ...data.author.node,
    }
  }

  // Clean up the categories to make them more easy to access
  if (data.categories) {
    modifiedData.categories = data.categories.edges.map(({ node }) => {
      return {
        ...node,
      };
    });
  }

  // Clean up the featured image to make them more easy to access
  if (data.featuredImage) {
    modifiedData.featuredImage = data.featuredImage.node;
  }

  if (data.tags) {
    modifiedData.tags = data.tags.edges.map(({ node }) => {
      return {
        name: node.name,
      };
    });
  }

  if(data.relatedPosts){
    modifiedData.relatedPosts = data.relatedPosts.map((post: IPostRaw) => {
      return mapPostData(post)
    })
  }

  // if(data.downloadManager?.downloads){
  //   modifiedData.downloadManager = data.downloadManager?.downloads.map(download => {
  //     return download.downloadDetails
  //   })
  // }

  if(data.tutorialManager?.paidProducts){
    const newPaidProducts = data.tutorialManager?.paidProducts.map(product => {
      const newProduct = {
        ...product,
        details: {
          ...product.details,
          licences: product.details.licences ? rearrangeLicenses(product.details.licences) : null,
        }
      }
      return newProduct
    })
    modifiedData.tutorialManager = {
      ...data.tutorialManager,
      paidProducts: newPaidProducts
    }
  }


  if(data.comments){
    modifiedData.comments = {
      pageInfo: data.comments.pageInfo,
      list: data.comments.edges.map(({ node }) => {
        return parseComment(node)
      }),
    }

  }

  return modifiedData

}

export function parseComment(node: IPostCommentRaw): IPostComment {

  return {
    ...node,
    replies: node.replies.edges.map(({ node }) => {
      return {
        ...node,
        parent: node.parent?.node.databaseId || null,
        author: node.author.node
      }
    }),
    parent: node.parent?.node.databaseId || null,
    author: node.author.node
  }
}

interface IMediaDetailSize {
  width: string
  file: string
  height: string
  name: string
  sourceUrl: string
}
export function getMediaSizeUrl(mediaSizes: IMediaDetailSize[] | undefined, name: string): IMediaDetailSize{
  if(mediaSizes === undefined || mediaSizes?.length < 1) return  {
    width: '',
    file: '',
    height: '',
    name: '',
    sourceUrl: ''
  }
  
  return mediaSizes.reduce((previousValue: any, currentValue: any) => {

    if(currentValue.name === name){
      return currentValue
    }else{ 
      return previousValue
    }

  }, {})
}

export function formatDate(date: string): string{
  const blogDate = new Date(date)
  const monthIndex: number = blogDate.getMonth()
  const day: number = blogDate.getDate()
  
  const months: string[] = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December'
  ]
  return `${months[monthIndex]} ${day}, ${blogDate.getFullYear()}`
}

export function splitProgramNameInTitle(title: string):{title: string, subTitle: string | undefined} {
  let newTitle = title
  switch (Boolean(title)) {

    case title.includes("in Procreate"):
      newTitle = title.replace("in Procreate", "")
      return {
        title: newTitle,
        subTitle: "in Procreate",
      }
      // return (
      //   <>
      //     <div className='mb-3'>{newTitle}</div>
      //     <div className={`${splitCss}`}>in Procreate</div>
      //   </>
      // )

    case title.includes("in Illustrator"):
      newTitle = title.replace("in Illustrator", "")
      return{
        title: newTitle,
        subTitle: "in Illustrator",
      }
      // return (
      //   <>
      //     <div className='mb-3'>{newTitle}</div>
      //     <div className={`${splitCss}`}>in Illustrator</div>
      //   </>
      // )
    case title.includes("in Adobe Illustrator"):
      newTitle = title.replace("in Adobe Illustrator", "")
      return {
        title: newTitle,
        subTitle: "in Adobe Illustrator",
      }
      // return (
      //   <>
      //     <div className='mb-3'>{newTitle}</div>
      //     <div className={`${splitCss}`}>in Adobe Illustrator</div>
      //   </>
      // )
    case title.includes("in InDesign"):
      newTitle = title.replace("in InDesign", "")
      return {
        title: newTitle,
        subTitle: "in InDesign",
      }
      // return (
      //   <>
      //     <div className='mb-3'>{newTitle}</div>
      //     <div className={`${splitCss}`}>in InDesign</div>
      //   </>
      // )
    case title.includes("in Photoshop"):
      newTitle = title.replace("in Photoshop", "")
      return {
        title: newTitle,
        subTitle: "in Photoshop",
      }
    default:
      return {
        title: newTitle,
        subTitle: undefined,
      }
  }
}

export function findString(str: string, arr: string[]) {
  return arr.find(item => item.includes(str))
}

export function findSkillLevel(categories: ICategories[]): ICategories | undefined {
  let tutorialsFound = false
  let defaultSkill = {
    name: "beginner",
  }
  let skillFoundStopLooking = false
  const skillFound = categories.reduce((previousValue: any, currentValue: ICategories):any => {
    if(currentValue.slug === "tutorials"){
      tutorialsFound = true
    }
    if(skillFoundStopLooking){
      return previousValue
    }
    switch(currentValue.slug ){
      
      case "advanced":
        skillFoundStopLooking = true
        return currentValue
      case "intermediate":
        skillFoundStopLooking = true
        return currentValue
      case "beginner":
        skillFoundStopLooking = true
        return currentValue
      default:
        return previousValue
    }
  }, {})

    return !isEmpty(skillFound) 
    ? skillFound 
    : isEmpty(skillFound) && tutorialsFound 
      ? defaultSkill
      : undefined
}

export function getLicense(licenses: ILicense[] | null, type: LicenseEnum) {
  if (!licenses) {
    return null
  }
  return licenses.reduce((acc, curr) => {
    if (curr.licenseType === type) {
      return curr
    }
    return acc
  }, {
    licenseType: '',
    price: 0,
    url: '',
  })
}