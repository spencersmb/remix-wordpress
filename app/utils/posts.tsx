import type { LicenseEnum } from "@App/enums/products";
import { isEmpty } from "lodash";

/**
 * @function flattenAllPosts
 * @tested - 6/7/2022
 * @description Go over all the posts and flatten the data to remove edges and nodes from the keys
 * 
 *
 * @returns mapped data array of posts
 **/
export function flattenAllPosts(posts: any): IPost[] | false {
  const postsFiltered = posts?.edges?.map(({ node = {} }) => node);
  return Array.isArray(postsFiltered) && postsFiltered.map(mapPostData)
}

/**
 * @function filterNodeFromTags
 * @tested - 6/7/2022
 * @description Remove Edges and nodes from the keys to return a clean array of tags
 * 
 *
 * @returns mapped array of tags
 **/
export function filterNodeFromTags(tags: IWpTags): Itag[] {
  return tags.edges.map(({ node }) => {
    return {
      name: node.name,
      slug: node.slug,
      count: node.count,
    };
  });
}

/**
 * @function rearrangeLicenses
 * @tested - 6/7/2022
 * @description Take in an array of licenses and rearrange them in specific order
 * Standard
 * Extended
 * Server
 * 
 *
 **/
export function rearrangeLicenses(licenses: ILicense[]): ILicense[] {

  return licenses.reduce((acc: any, license) => {

    // Add it to the beginning of the array
    if (license.licenseType === 'standard') {
      acc.unshift(license)
    }

    if (license.licenseType === 'extended') {

      // check if the array has an item in it and make sure its not the standard license because standard is supposed to always be first
      if (acc.length === 1 && acc[0].licenseType !== 'standard') {
        acc.unshift(license) // means that the item in the array is the server license and we need to add the extended license before it

        // if the array has 2 items in it already it means we need to put the extended license in the middle of the array
      } else if (acc.length === 2) {
        const standard = acc[0]
        const server = acc[1]
        acc = [
          standard,
          license,
          server
        ]
      } else {
        acc.push(license)
      }
    }

    // always add server to the end of the array
    if (license.licenseType === 'server') {
      acc.push(license)
    }

    return acc
  }, [])
}

/**
 * @function mapPostData
 * @tested - 6/7/2022
 * @description Map over POST data, filter and return a clean object
 *
 **/
export function mapPostData(post: IPostRaw | {} = {}): IPost {
  const data = { ...post };
  let modifiedData: any = { ...post }

  // Clean up the author object to avoid someone having to look an extra
  // level deeper into the node
  if (data.author) {
    modifiedData.author = {
      ...data.author.node,
    }
  }

  // Clean up the categories to make them more easy to access
  if (data.categories) {
    modifiedData.categories = filterCategories(data.categories.edges);
  }

  // Clean up the featured image to make them more easy to access
  if (data.featuredImage) {
    modifiedData.featuredImage = data.featuredImage.node;
  }

  if (data.tags) {
    modifiedData.tags = data.tags.edges.map(({ node }) => {
      return {
        name: node.name,
        slug: node.slug,
        count: node.count,
      };
    });
  }

  if (data.relatedPosts) {
    modifiedData.relatedPosts = data.relatedPosts.map((post: IPostRaw) => {
      return mapPostData(post)
    })
  }

  if (data.tutorialManager) {
    modifiedData.tutorialManager = {
      ...data.tutorialManager,
      resources: data.tutorialManager.resources ? mapPostResources(data.tutorialManager.resources) : []
    }
  }

  if (data.comments) {
    modifiedData.comments = {
      pageInfo: data.comments.pageInfo,
      list: data.comments.edges.map(({ node }) => {
        return parseComment(node)
      }),
    }

  }

  return modifiedData

}

/**
 * @function parseComment
 * @tested - 6/7/2022
 * @description Return a filtered clean comment object
 * 
 *
 **/
export function parseComment(node: IPostCommentRaw): IPostComment {

  return {
    ...node,
    replies: node.replies.edges.map(({ node }) => {
      return {
        ...node,
        parent: node.parent?.node.databaseId || null,
        author: node.author.node
      }
    }),
    parent: node.parent?.node.databaseId || null,
    author: node.author.node
  }
}

/**
 * @function formatDate
 * @tested - 6/7/2022
 * @description Return a readable date from a WP server timestamp
 * 
 *
 **/
export function formatDate(date: string): string {
  const blogDate = new Date(date)
  const monthIndex: number = blogDate.getMonth()
  const day: number = blogDate.getDate()

  const months: string[] = [
    'Jan',
    'Feb',
    'March',
    'April',
    'May',
    'June',
    'July',
    'Aug',
    'Sept',
    'Oct',
    'Nov',
    'Dec'
  ]
  return `${months[monthIndex]} ${day}, ${blogDate.getFullYear()}`
}

/**
 * @function parseStringForSpecialCharacters
 * @tested - 6/7/2022
 * @description Remove certain characters from a string
 * Used on Blog Post titles
 * 
 *
 **/
export function parseStringForSpecialCharacters(string: string): string {
  return string.replace(/&/g, '').replace(/</g, '').replace(/>/g, '').replace(/"/g, '').replace(/'/g, '').replace(/\+/g, '')
  // return string.replace(/\+/g, '');
}

/**
 * @function checkTitleForBrackets
 * @tested - 6/7/2022
 * @description If brackets are in the title, remove them and return the title, then add the text that was in the brackets as a subtitle.
 * 
 *
 **/
export function checkTitleForBrackets(title: string): { title: string, subTitle: string | undefined } {
  let sqaureBrackets = title.substring(
    title.indexOf("[") + 1,
    title.lastIndexOf("]")
  );

  let roundBrackets = title.substring(
    title.indexOf("(") + 1,
    title.lastIndexOf(")")
  );

  if (sqaureBrackets.length > 0) {
    let newtTitle = title.replace(`[${sqaureBrackets}]`, '')
    return {
      title: newtTitle,
      subTitle: parseStringForSpecialCharacters(sqaureBrackets)
    }
  }

  if (roundBrackets.length > 0) {
    let newtTitle = title.replace(`(${roundBrackets})`, '')
    return {
      title: newtTitle,
      subTitle: parseStringForSpecialCharacters(roundBrackets)
    }
  }

  return {
    title: title,
    subTitle: undefined
  }
}

/**
 * @function splitProgramNameInTitle
 * @tested - 6/7/2022
 * @description If certain key software words are in the title, remove them and return the title, then add the text as a subtitle.
 * 
 *
 **/
export function splitProgramNameInTitle(title: string): { title: string, subTitle: string | undefined } {
  let newTitle = title
  switch (Boolean(title)) {

    case title.includes("in Procreate"):
      newTitle = title.replace("in Procreate", "")
      return {
        title: newTitle,
        subTitle: "in Procreate",
      }

    case title.includes("in Illustrator"):
      newTitle = title.replace("in Illustrator", "")
      return {
        title: newTitle,
        subTitle: "in Illustrator",
      }
    case title.includes("in Adobe Illustrator"):
      newTitle = title.replace("in Adobe Illustrator", "")
      return {
        title: newTitle,
        subTitle: "in Adobe Illustrator",
      }
    case title.includes("in InDesign"):
      newTitle = title.replace("in InDesign", "")
      return {
        title: newTitle,
        subTitle: "in InDesign",
      }
    case title.includes("in Photoshop"):
      newTitle = title.replace("in Photoshop", "")
      return {
        title: newTitle,
        subTitle: "in Photoshop",
      }
    default:
      return {
        title: newTitle,
        subTitle: undefined,
      }
  }
}

/**
 * @function findString
 * @tested - 6/7/2022
 * @description Return a string that is in the array if found, otherwise is undefined
 * 
 *
 **/
export function findString(str: string, arr: string[]): string | undefined {
  return arr.find(item => item.includes(str))
}

/**
 * @function findSkillLevel
 * @tested - 6/7/2022
 * @description This function takes the blog catgegories and looks for skill level. Will return the skill level if found from highest to lowest in-case there are multiple skill levels.
 * 
 *
 **/
export function findSkillLevel(categories: ICategories[]): ICategories | undefined {
  let tutorialsFound = false
  let defaultSkill: ICategories = {
    name: 'beginner',
    databaseId: 0,
    slug: 'beginner',
    id: 'beginner',
  }

  // grab the items with skill levels in them
  const skillsFound = categories.reduce((previousValue: any, currentValue: ICategories, index: number): any => {
    let level

    switch (currentValue.slug) {
      case "tutorials":
        tutorialsFound = true
        return previousValue
      case "advanced":
        level = 3
        previousValue.push({
          cat: currentValue,
          level: level,
          index
        })
        return previousValue
      case "intermediate":
        level = 2
        previousValue.push({
          cat: currentValue,
          level: level,
          index
        })
        return previousValue
      case "beginner":
        level = 1
        previousValue.push({
          cat: currentValue,
          level: level,
          index
        })
        return previousValue
      default:
        return previousValue
    }

  }, [])

  // Return default skill if tutorial is found but forgot to add skill level in WP backend.
  if (skillsFound.length === 0 && tutorialsFound) {
    return defaultSkill
  } else if (skillsFound.length === 0 && !tutorialsFound) {
    return undefined
  }

  // find the highest skill level
  return skillsFound.reduce((previousValue: any, currentValue: any): any => {
    if (currentValue.level > previousValue.level) {
      return currentValue
    }
    return previousValue
  }).cat
}

/**
 * @function getLicense
 * @tested - 6/7/2022
 * @description Search the License array for a specific license based on the License Enum
 * 
 *
 **/
export function getLicense(licenses: ILicense[] | null, type: LicenseEnum): ILicense | null {
  if (!licenses) {
    return null
  }

  const foundLicense = licenses.reduce((acc: any, curr: ILicense) => {
    if (curr.licenseType === type) {
      return curr
    }
    return acc
  }, {})

  if (isEmpty(foundLicense)) {
    return null
  }

  return foundLicense
}

/**
 * @function flattenAllCourses
 * @tested - 6/7/2022
 * @description Go over all the courses and flatten the data to remove edges and nodes from the keys
 * 
 *
 **/
export function flattenAllCourses(courses: ICoursesRaw): ICourse[] | false {
  const coursesFiltered = courses?.edges?.map(({ node = {} }) => node);
  return Array.isArray(coursesFiltered) && coursesFiltered.map(mapCourseData)
}

/**
 * @function mapCourseData
 * @tested - 6/7/2022
 * @description Map over Course data, filter and return a clean object
 * 
 *
 **/
export function mapCourseData(course: ICourseRaw | {} = {}): ICourse {
  const data = { ...course };
  let modifiedData: any = { ...course }

  if (data.featuredImage) {
    modifiedData.featuredImage = data.featuredImage.node;
  }

  if (data.details?.courseTags && data.details.courseTags.length > 0) {
    modifiedData.details.courseTags = data.details.courseTags.map(item => {
      return item.tag
    })
  }

  return modifiedData
}

/**
 * @function filterCategories
 * @tested - 6/29/2022
 * @description Map over Categories and return a clean object
 * 
 *
 **/
export function filterCategories(categories: ICategoryRaw[]): ICategories[] {
  return categories.map(({ node }) => {
    return {
      ...node,
    };
  })
}

/**
 * @function removeLastItemFromArray
 * @tested - 6/29/2022
 * @description Removes the last item from an array and return the new array AND the last item both as seperate objects to use in the breadcrumbs component
 * 
 *
 **/
export function removeLastItemFromArray(array: any[] | undefined) {
  if (!array) {
    return { lastElement: null, modifiedArray: null }
  }
  const newArray = [...array]
  const lastElement = newArray.pop()
  return {
    modifiedArray: newArray,
    lastElement
  }
}

export enum POST_RESOURCE_ENUMS {
  PRODUCT = 'product',
  COURSE = 'course',
  DOWNLOAD = 'download',
  SWATCH = 'colorSwatch'
}

/**
 * 
 * @function getResource 
 * @tested - 08/22/2022 
 */
export function getResource({ resources, resourceName }: { resources: IPostResource[], resourceName: POST_RESOURCE_ENUMS }): IPostResource | null {
  let foundResource = null
  resources.forEach(resource => {
    const keys = Object.keys(resource)
    if (keys.includes(resourceName)) {
      foundResource = resource
    }
  })
  return foundResource
}

/**
 * 
 * @function mapPostResources 
 * @tested - 08/22/2022 
 */
export function mapPostResources(resources: IPostResource[]) {
  return resources.map((resource, index) => {
    const keys = Object.keys(resource)

    if (keys.includes('colorSwatch')) {
      return resource
    }

    if (keys.includes('course')) {
      return resource
    }

    if (keys.includes('product')) {
      if (!resource.product) {
        return resource
      }
      return {
        ...resource,
        product: {
          ...resource.product,
          productDetails: {
            ...resource.product.productDetails,
            licences: resource.product.productDetails.licences ? rearrangeLicenses(resource.product.productDetails.licences) : null,
          }
        }
      }
    }

    return resource
  })
}


export function reducePostResourceData(resources: IPostResource[]) {
  return resources.reduce((acc, resource) => {
    const keys = Object.keys(resource)

    if (keys.includes('colorSwatch')) {
      return {
        ...acc,
        colorSwatch: resource.colorSwatch
      }
    }

    if (keys.includes('course')) {
      if (!resource.course) {
        return acc
      }
      return {
        ...acc,
        course: {
          ...resource.course,
          description: resource.description
        }
      }
    }

    if (keys.includes('product')) {
      if (!resource.product) {
        return acc
      }
      return {
        ...acc,
        product: {
          ...resource.product,
          description: resource.description,
          productDetails: {
            ...resource.product.productDetails,
            licences: resource.product.productDetails.licences ? rearrangeLicenses(resource.product.productDetails.licences) : null,
          }
        }
      }
    }

    if (keys.includes('download')) {
      return {
        ...acc,
        download: resource.download
      }
    }

    return acc
  }, {})
}

// export function createThumbnailImage(
//   tutorialManager: ITutorialManager,
//   defaultSource: IMediaDetailSize,
//   title: string,
//   featuredPost: boolean = false
// ) {

//   const defaultImage = () => (
//     // <div className={classNames(featuredPost ? 'rounded-2.5xl' : 'mb-8 h-[250px]', "default_image relative overflow-hidden")}>
//     <div className={classNames(featuredPost ? 'rounded-2.5xl' : 'mb-8', "default_image relative overflow-hidden")}>
//       {defaultSource.sourceUrl.length !== 0 &&
//         // <img className={classNames(featuredPost ? '' : 'absolute max-w-none top-[50%] w-full translate-y-[-50%]', "")}
//         <img className={classNames(featuredPost ? '' : '', "")}
//           src={defaultSource.sourceUrl} alt={title} />}
//     </div>
//   )
//   if (!tutorialManager.thumbnail || !tutorialManager.thumbnail.image) {
//     return defaultImage()
//   }

//   switch (tutorialManager.thumbnail.type) {
//     case "make":
//       return (
//         <>
//           <div className={classNames(featuredPost
//             ? `absolute top-[-20px] left-[15px] w-[45%] rotate-[352deg]`
//             : `absolute top-[10px] left-[15px] w-[40%]`, 'z-10')}>
//             <img className="relative z-10" src="/images/make-this.png" alt={`Make this tutorial: ${title}`} />
//             <span className={classNames(featuredPost
//               ? 'top-[45px]'
//               : 'top-[20px] ',
//               'absolute w-[40%] left-[71%] z-[5]')}>
//               <img src="/images/make-this-arrow-1.png" alt={`Make this tutorial: ${title}`} />
//             </span>
//           </div>
//           {!featuredPost && <div className="absolute top-[15px] right-[30px] w-[25%] opacity-70">
//             <img src="/images/video-tutorial-text.png" alt={`Video tutorial: ${title}`} />
//           </div>}
//           <div className="relative">
//             <div className="rounded-2.5xl overflow-hidden">
//               <img src={tutorialManager.thumbnail.image.sourceUrl} alt={`${tutorialManager.thumbnail.image.altText} Main Image`} />
//             </div>
//             {tutorialManager.colorPalette && <div style={{ backgroundColor: tutorialManager.colorPalette.iconBackgroundColor }} className="absolute rounded-full bottom-[-10px] tablet:bottom-[-10%] right-[30px] w-[100px] h-[100px] laptop:bottom-[-6%] desktop:w-[138px] desktop:h-[138px] bg-slate-500 desktop:top-auto desktop:bottom-[-10px] flex justify-center items-center">
//               <span style={{ color: tutorialManager.colorPalette.iconTextColor }} className="transform rotate-[-8deg] text-center font-sentinel__SemiBoldItal tablet:leading-4 desktop:text-xl desktop:leading-6">
//                 Free Color Swatches
//               </span>
//             </div>}
//           </div>

//         </>
//       )
//     default:
//       return defaultImage()
//   }
// }


// export function findSkillLevelDepricated(categories: ICategories[]): { name: string } | undefined {
//   let tutorialsFound = false
//   let defaultSkill = {
//     name: "beginner",
//   }
//   let skillFoundStopLooking = false

//   const skillFound = categories.reduce((previousValue: any, currentValue: ICategories): any => {
//     if (currentValue.slug === "tutorials") {
//       tutorialsFound = true
//     }
//     if (skillFoundStopLooking) {
//       return previousValue
//     }

//     switch (currentValue.slug) {

//       case "advanced":
//         skillFoundStopLooking = true
//         return currentValue
//       case "intermediate":
//         skillFoundStopLooking = true
//         return currentValue
//       case "beginner":
//         skillFoundStopLooking = true
//         return currentValue
//       default:
//         return previousValue
//     }
//   }, {})

//   return !isEmpty(skillFound)
//     ? skillFound
//     : isEmpty(skillFound) && tutorialsFound
//       ? defaultSkill
//       : undefined
// }
