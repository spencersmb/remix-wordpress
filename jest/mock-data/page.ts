import { mockFeaturedImage } from "./images";
import { mockAuthorData, mockSeo } from "./posts";

export const mockPage: IPage = {
  seo: {
    title: "Page Title",
    metaDesc: "Page description",
    opengraphModifiedTime: "2020-01-01T00:00:00.000Z",
    opengraphPublishedTime: "2020-01-01T00:00:00.000Z",
    readingTime: 1
  },
  title: "Page Title",
  slug: "page-slug",
  content: "Page Content",
  author: mockAuthorData,
  featuredImage: mockFeaturedImage,
  date: "2020-01-01",
  id: "page-id",
}