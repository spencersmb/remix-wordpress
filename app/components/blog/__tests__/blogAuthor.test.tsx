import { mockPostData } from '@TestUtils/mock-data/posts';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import BlogAuthor from '../blogAuthor';
describe('Blog Author Tests', () => {

  it('Should render Blog Author Component ', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <BlogAuthor post={mockPostData} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  })

})