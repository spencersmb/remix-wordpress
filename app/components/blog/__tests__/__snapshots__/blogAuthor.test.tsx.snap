// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog Author Tests Should render Blog Author Component  1`] = `
<div
  className="flex flex-col flex-wrap items-center tablet:flex-row"
  data-testid="blogAuthor-test"
>
  <div
    className="tablet:mr-2"
  >
    <div
      className="w-[100px]"
    >
      <div
        className="overflow-hidden rounded-full max-w-none lazy-load-wrapper lazy-load-wrapper-block lazy-load-image-full"
      >
        <span
          className=" lazy-load-image-background blur"
          style={
            Object {
              "backgroundImage": "",
              "backgroundSize": "",
              "color": "transparent",
              "display": "inline-block",
              "height": "96px",
              "width": "96px",
            }
          }
        >
          <span
            className=""
            style={
              Object {
                "display": "inline-block",
                "height": "96px",
                "width": "96px",
              }
            }
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="my-4 flex flex-col justify-center flex-[1] text-center tablet:my-0 tablet:text-left"
  >
    <div
      className="mb-2 text-sm text-sagae-700"
    >
      Written By
    </div>
    <div
      className="text-primary-600 font-sentinel__SemiBoldItal text-h3"
    >
      Teela Cunningham
    </div>
    <div
      className="text-base text-primary-600"
    >
      Every Tuesday's content creator and founder.
    </div>
  </div>
  <div
    className="mt-2 flex-[1_1_100%] tablet:flex-[0_1_auto] items-center justify-center tablet:self-end pb-2 tablet:mt-0"
  >
    <a
      className="btn btn-outline"
      href="/about"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
    >
      About Me
    </a>
  </div>
</div>
`;
